### ISSUE TITLE
Cone beam shows a dip artifact in the center

### ISSUE THREAD

**Issue by fangq** [fangq/mcx#86]
This issue was reported by Yueqi Guo <guoyueqi912 at gmail> in the below mailing list discussion

https://groups.google.com/forum/?hl=en#!topic/mcx-users/kmY2Obd4iZg

basically, for cone beam, a small dip was observed at the center of the beam, where relatively uniform distribution id expected.

On 1/17/20 11:09 AM, Yueqi Guo wrote:
> Hi Dr. Fang,
>
> Thank you for answering all my previous questions! I really appreciate it.
> I have two more questions ...
> 1) how does the mcxlab initiated each photon for a "cone" light source with NA=0.03? I assume the initial angle is some sort of uniform random distribution? I focused the cone beam 500um above the medium surface by just setting the first 500um of medium as "cfg.vol = 0". On the fluence map, I always see a small indentation in the middle of the medium surface, do you know why is that? (attached figure)
>
> 2) how is the photon position related to the index of voxels (for fluence map and cfg.vol etc.)? For example, for a voxel with index "100, 100, 1", does it include photons with the position "[99.5, 100.5), [99.5, 100.5), [0.5, 1.5)"? Or something like   "[100, 101), [100, 101), [1, 2)"
>

![Annotation 2020-01-17 110721](https://user-images.githubusercontent.com/226913/72846484-ade8b680-3c6e-11ea-8f04-b366f7b71bf3.png)

**Comment by fangq** [comment]
This issue was confirmed using the below test code

On 1/21/20 4:56 PM, Qianqian Fang wrote:
>
> you were actually right! I was able to reproduce the artifact you saw (small dip in the middle of the cone beam, ~5% lower) using the below code, my plots are attached
>
>
>     clear cfg;
>     cfg.nphoton=1e8;
>     cfg.vol=uint8(ones(60,60,60));
>     cfg.gpuid=1;
>     cfg.autopilot=1;
>     cfg.prop=[0 0 1 1;0.005 1 0.8 1.37];
>     cfg.tstart=0;
>     cfg.seed=99999;
>
>     % a cone beam
>     cfg.srctype='cone';
>     cfg.unitinmm=0.005;
>     cfg.srcpos=[30 30 -500];
>     cfg.srcdir=[0 0 1];
>     cfg.tend=1e-11;
>     cfg.tstep=1e-11;
>     cfg.srcparam1=[0.03 0 0 0];
>     cfg.srcparam2=[0 0 0 0];
>     flux=mcxlab(cfg);
>     fcw=flux.data*cfg.tstep;
>     subplot(121);
>     imagesc(log10(abs(squeeze(fcw(:,:,1)))))
>     axis equal; colorbar
>     title('a uniform cone beam (uniform solid-angle)');
>
>     subplot(122);
>     plot(abs(squeeze(fcw(:,30,1))))
>
>
> the above code does something similar to yours, except that I increased the spacing from 100 voxels to 500 voxels to zoom-in the beam profile.
>
> I agree that this dip is undesired, and should be fixed - I have a feeling that this is due to the way how I generated the random zenith angle, instead of [0, pi), I think the code generates (eps, pi) where eps is around 1e-6 for single-precision number.
>
> I just create a github ticket to document and track this problem
>
> https://github.com/fangq/mcx/issues/86
>
> thanks for reporting this! I will let you know if I have a fix committed.
>
> Qianqian

![cone_beam_dip_artifact](https://user-images.githubusercontent.com/226913/72846893-79c1c580-3c6f-11ea-9149-848c9a09e4fd.png)

**Comment by fangq** [comment]
@YGUO29: I am happy to let you know that the problem has been fixed. The updated fluence profile plot is attached.

this is what happened: your NA is quite small (~1e-3), at such a small angle, the numerical round-off error (EPS of single-precision is around 1e-7) when converting the random number to a random angle via `arccos` becomes noticeable at the small zenith angle (i.e. near the center of the cone), as a result of the single-precision math in the mcx GPU kernel.

another problem is related to the not-so-efficient way that I used to create the random cone beam angle - I first generate a random angle on a full sphere, and only launch the photon if the zenith angle is within your cone half-angle (srcparam1(1)). If an angle is bigger than the cone, I will repeat the random angle calculation until the condition is satisfied. This becomes increasingly inefficient when you cone angle becomes smaller - in such case, one has to compute many random angles to get one that is small enough. 

to fix these problems, I made two changes

1. I changed the way to generate random cone beam - and made it possible to create the cone angle without needing to repeat (should have done this long time ago), 
2. but the above change alone s not enough to fix the dip (actually it became a small bump if I still use single-precision), so I had to change the angle calculations to double-precision.

after both changes, the uneven angular distribution is fixed, in the meantime, mcx simulation speed for this small cone case was accelerated by 9x! (from 1676 photon/ms to 14859 photon/ms on a TitanV). I wish I had done this long time ago!

In the past, I tried to avoid double precision (fp64) at all cost, because most consumer GPU does not have good fp64 hardware, but I tested on both Titan V (which has good fp64) and GTX1080 (which has very weak fp64 support), the speed impact due to the single->double change was very small.

let me know if you see any addition issues with this fix.

![fix_issue_86](https://user-images.githubusercontent.com/226913/73327307-64c2d480-4223-11ea-9aed-db1b6d812c62.png)
